[IV-24-25] Objetivo 2

pablotl0 opened the pull request:
Sobre la estructura del repositorio

* [x] ¿He seguido las mejores prácticas en el nombre de las clases y ficheros y
   disposición de los mismos?
* [x] ¿He mirado tanto reglas generales (para cualquier repo) como reglas
específicas para el lenguaje con el que se está trabajando?

Sobre el planteamiento

* [x] ¿Se han planteado una serie de issues, asignados a las historias de
usuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e
identifiquen las diferentes partes del problema?
* [x] ¿Los issues representan un problema, y no una tarea?

Sobre el análisis del problema

* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
   que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
   una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo
   objeto valor que no dependa de ningún otro (y que sea la base de muchos
   otros)?.

Sobre el código

* [x] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están relacionados con a una HU?
* [x] ¿Ha asignado el propietario/la propietaria del repositorio todos los
issues planteados al *milestone* correspondiente?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Es el milestone sobre el que estoy trabajando un producto
   mínimamente viable? ¿O tengo que solicitar al product manager que precise de
   qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente
correcto?

pablotl0 committed:
closes #11, creando iv.yaml con el lenguaje acordado

pablotl0 committed:
closes #14,
creo el objeto valor Producto con el objetivo de representar un artículo. El objeto tiene un atributo 'nombre' para representar de manera básica el producto.

pablotl0 committed:
closes #13,
creo la entidad ElementoLista con el objetivo de representar los componentes de la lista de deseos. El objeto tiene un atributo 'producto' para representar de manera básica los componentes.

pablotl0 committed:
closes #12,
creo la entidad ListaDeseos con el objetivo de representar los regalos que un usuario quiere recibir. El objeto tiene un atributo 'deseos' para representar de manera básica el conjunto de regalos.

pablotl0 committed:
closes #16,
creo la entidad Usuario con el objetivo de representar unla persona que quiere recibir regalos. El objeto tiene unos atributos 'id', 'nombre', 'listaDeseos'  para representar de manera básica al usuario.

pablotl0 committed:
Fix #14

En relación a https://github.com/evaanngiil/WishfulGiving/pull/17#issuecomment-2480564420 modifico la clase Producto para añadirle un precio al producto

pablotl0 committed:
Fix #11

pablotl0 committed:
Fix #11

evaanngiil commented:
Por mi parte considero que todo está correcto, siguiendo la definición de modelo y sus características.

pablotl0 committed:
Update iv.yaml

pablotl0 committed:
Rename Usuario.java to usuario.java

pablotl0 commented:
@JJ Listo para revisión

evaanngiil commented:
* [x] ¿Los issues que ha planteado la persona son efectivamente problemas y
evidencian comprensión de las historias de usuario? Si no es así, es mejor que
se le transmita antes de que empiece a crear cualquier código erróneo.
* [x] ¿Has comprobado que efectivamente el compañero/a ha seguido *su* lista de
      comprobación?
* [x] ¿El código creado es una modelización válida de las historias de usuario
      creadas por ti, y te sirve como base para programar la lógica de negocio?
* [x] ¿El código creado sigue las mejores prácticas para el lenguaje de
      programación que hayáis consensuado, tanto en identificadores, como
      nombres de ficheros, como disposición de los diferentes ficheros en el
      repositorio? Por ejemplo, ¿todos los commits se refieren a un issue y
      aportan código para resolver el problema planteado en el mismo?
* [x] ¿He comprobado que el código compile, o que quien lo ha programado ha lo
ha hecho y documentado en el repositorio?
* [x] ¿He añadido el PR y todos los issues que se resuelvan en él al milestone
correspondiente, que será el primero?
* [x] ¿Cubre totalmente el milestone 0, o hará falta modelización adicional en
el objetivo 4 para poder programar la lógica de negocio?
* [x] En resumen, ¿considero que con lo hecho puedo ponerme a trabajar
directamente en el siguiente milestone, que incluye la lógica de negocio?

pablotl0 commented:
@JJ listo para revision

JJ commented:
> Por mi parte considero que todo está correcto, siguiendo la definición de modelo y sus características.

Evan, es prácticamente imposible que esté todo correcto a la primera. Ni siquiera los profesionales que trabajan en una empresa lo consiguen. Por favor, mira bien lo que hace, mira bien los errores frecuentes (que hay un buen montón) y vuelve a revisar.

pablotl0 committed:
fix #13,
elimino la clase ElementoLista ya que no aporta ninguna información adicional haciendo innecesaria esta clase y se puede sustituir por la clase Producto

pablotl0 committed:
fix #12,
he modificado el tipo del ArrayList a Producto al eliminar la clase ElementoLista por no aportar valor sobre el producto

pablotl0 committed:
fix #12,
he arreglado el constructor de la clase ListaDeseos añadiendo el parametro id

pablotl0 commented:
@JJ listo para revision

pablotl0 committed:
fix #12,
elimino el id de la clase al ser innecesario para gestionar la lista de elementos

pablotl0 committed:
fix #16,
elimino el id de la clase al ser innecesario para identificar un usuario podiendo ser representado este por su nombre

pablotl0 commented:
@JJ listo para revisión.

pablotl0 committed:
Fix #16

Arreglo un error sobre el id

evaanngiil reviewed (approved):
El modelo resuelve los requerimientos de la HU asociada al milestone 0. Además, el objeto valor esta separado de la entidad. Lo doy por aprobado

JJ commented on src/usuario.java:
```diff
@@ -0,0 +1,21 @@
+public class Usuario { 
```
¿Y quién maneja los diferentes usuarios?
¿Necesitarás una entidad?

JJ replied:
No se cierran *nunca* los comentarios sin decir por qué.
Si en este caso la respuesta es #16 y a "como se manejan los usuarios" creas un "manejador de usuarios" es que no has entendido demasiado bien la pregunta ni estás haciendo un modelo adecuado.

pablotl0 replied:
@JJ La entidad me sirve para administrar los usuarios,  asegurando de que la lógica se maneje de manera centralizada.  Ya que el HashMap solo almacena datos y no define cómo se usan ni gestiona la lógica de los datos.

JJ commented on src/ListaDeseos.java:
```diff
@@ -0,0 +1,10 @@
+import java.util.ArrayList;
+
+public class ListaDeseos {
+    private ArrayList<Producto> deseos;
```
¿Qué valor añade al cliente ponerle un nombre diferente a un array?

JJ reviewed (changes requested):
Por favor, consulta los errores comunes también.

pablotl0 committed:
#12,
elimino la clase ListaDeseos ya que no aporta valor adicional al array de deseos

pablotl0 committed:
#12,
 modifico la clase sustituyendo el objeto ListaDeseos por el vector de deseos

pablotl0 committed:
#16: Crear una clase de gestión de usuarios centralizada para mejorar la coherencia y la escalabilidad del sistema.
Esta clase es necesaria para consolidar la lógica de gestión de usuarios y permitir una administración uniforme
y centralizada, lo que facilitará futuras modificaciones y la integración con otros módulos.

pablotl0 commented:
@evaanngiil listo para revision

evaanngiil reviewed (changes requested):
El objeto valor `Producto` es manejado por la entidad `Usuario`, a su vez esta es manejada por la entidad principal `GestorDeUsuarios`. Sin embargo, creo que si la entidad se llama `GestorDeUsuarios` no tiene mucho sentido llamar al archivo _sistema_

pablotl0 committed:
Rename sistema.java to gestorDeUsuarios.java

pablotl0 committed:
#11

cambio la entidad a gestorDeUsuarios porque es esta clase quien controla los objetos valores y por tanto la entidad.

pablotl0 commented:
@evaanngiil ya he corregido el error.

evaanngiil reviewed (changes requested):
Recuerda también seguir las buenas prácticas de programación usando _snake_case_ o _camelCase_ (usa el que creas conveniente) en el renombrado de los archivos de Java. Lo ideal sería no mezclar ambos tipos de convenciones

pablotl0 committed:
Rename Producto.java to producto.java

pablotl0 commented:
@evaanngiil ya he cambiado el nombre de los archivos

pablotl0 committed:
#16 Cambio el array por un hashamp para mejorar el manejo de objetos

pablotl0 commented:
@evaanngiil he cambiado la estructura del gestor de Usuarios para facilitar la búsqueda de estos. En relación con los errores comunes.

pablotl0 commented:
@JJ listo para revision

JJ commented:
> Recuerda también seguir las buenas prácticas de programación usando _snake_case_ o _camelCase_ (usa el que creas conveniente) en el renombrado de los archivos de Java. Lo ideal sería no mezclar ambos tipos de convenciones

En realidad, cada lenguaje de programación tiene su propia convención. No puedes usar "el que creas conveniente" (salvo que en Java sea así)

JJ commented on src/gestorDeUsuarios.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.HashMap;
+
+public class GestorDeUsuarios { 
+    private HashMap<String, Usuario> usuarios; // Cambiado a HashMap
```
No se usan comentarios en código limpio y menos para indicar qué se ha hecho. Para eso están los mensajes de commit.

JJ replied:
¿Qué es la clave aquí?

pablotl0 replied:
@JJ La clave es un String refiriéndose al nombre del usuario.

JJ commented on src/gestorDeUsuarios.java:
```diff
@@ -0,0 +1,9 @@
+public class GestorDeUsuarios { 
+    private HashMap<String, Usuario> usuarios; // Cambiado a HashMap
+
+    public GestorDeUsuarios() {
```
Tener que ir al commit específico para ver qué se hace no ayuda a revisar. En la primera línea del mensaje de commit tiene que haber algo más que el número del issue.

pablotl0 replied:
@JJ He modificado el mensaje del commit 900f1550. 
Explicando que es necesario una entidad para gestionar la lógica del conjunto de usuarios

JJ replied:
El comentario sigue sin resolver y sigue sin estar justificada la existencia de esta clase específica.

pablotl0 replied:
@JJ En el mensaje del commit 8cfe218 se explica porque debe existir la clase

JJ reviewed (changes requested):
El principal el problema aquí es no resolver bien el comentario, sin modificar el código sobre el que se comenta o decir en qué commit se hace. Pero el problema mayor es aprobar el PR sin entender muy bien qué es lo que hace y por qué.

pablotl0 committed:
#16 Elimino comentarios

pablotl0 commented:
@JJ revision

JJ commented on src/producto.java:
```diff
@@ -0,0 +1,21 @@
+    
+    public Producto(String nombre, double precio) {
+        validarProducto(nombre, precio);
+        this.nombre = nombre.trim();
```
¿Por qué haces `trim`?

JJ commented on src/usuario.java:
```diff
@@ -0,0 +1,19 @@
+
+public class Usuario { 
+    private String nombre;
+    private ArrayList<Producto> listaDeseos;
```
@evaanngiil ¿esto cubre la historia de usuario?

evaanngiil replied:
Al tratarse de un usuario que va a recibir regalos creo que faltaría información para cubrir la lógica de negocio y resolver la HU #4. ¿Qué pasaría si pido unas zapatillas o una prenda de ropa? Con la información que contiene `Producto` no bastaría para resolver el problema

pablotl0 replied:
@JJ yo puedo añadir mas atributos pero si en la historia de usuario no se referencian esas necesidades entonces no es un producto mínimo ya que me estoy inventando atributos. ¿Qué atributos son los necesarios realmente?

JJ replied:
Voy a comentar en la HU

JJ replied:
Por otro lado, con esta estructura de datos no garantizas una fuente única de verdad para los productos.

JJ reviewed (changes requested):
Aparte de tener que resolver *todos* los comentarios, una vez más @evaanngiil tendría que replantearse si el modelo es correcto y suficiente para programar la lógica de negocio.

pablotl0 committed:
#14 Elimino el metodo trim.
Al ser innecesaro para la validacion.

pablotl0 commented:
@JJ revision

pablotl0 committed:
elimino las clases para empezar de nuevo de acuerdo a la nueva HU

pablotl0 committed:
closes #14 define un regalo con atributos esenciales: nombre y precioson los atributos porque definen cual es el regalo y cuánto costará, respectivamente.

pablotl0 committed:
closes #23 el presupuesto se modela como un valor total y un monto restante. El presupuesto es fijo para garantizar que se mantenga una referencia clara y consistente, se prevengan errores y se facilitela gestión del dinero de forma controlada

pablotl0 committed:
closes #12, esta clase relaciona destinatario, presupuesto y regalos cada uno con una prioridad. Esta clase integra todas las entidades para modelar el dominio completo, asegurando que todas las interacciones dentro de la lista de regalos se gestionen de manera coherente y eficiente.

pablotl0 committed:
#12 Arreglo el nombre del fichero para que coincida con la clase

pablotl0 committed:
#12: Creo una clase de gestión de listas de regalos centralizada para mejorar la coherencia y la escalabilidad del sistema.
Esta clase es necesaria para consolidar la lógica de gestión de liastas de regalos y permitir una administración uniforme
y centralizada, lo que facilitará futuras modificaciones y la integración con otros módulos.

pablotl0 commented:
@evaanngiil revision

evaanngiil commented on src/Presupuesto.java:
```diff
@@ -0,0 +1,13 @@
+public class Presupuesto {
+    private final double presupuestoTotal; 
+    private double presupuestoRestante; 
```
La clase Presupuesto no se va a utilizar en ninguna otra ocasión por lo que no sé porqué se encuentra separado. Además, no veo necesario crear una clase para un presupuesto el cual puede ser simplemente un valor flotante.

evaanngiil replied:
La clase Presupuesto no se va a utilizar en ninguna otra ocasión por lo que no sé porqué se encuentra separado. Además, no veo necesario crear una clase para un presupuesto el cual puede ser simplemente un valor flotante.

evaanngiil commented on src/ListaRegalos.java:
```diff
@@ -0,0 +1,35 @@
+        this.regalos = new ArrayList<>();
+    }
+
+    public static class ElementoLista {
+        private final Regalo regalo;
+        private final int prioridad;
```
Un elemento en la lista será un regalo con su prioridad asociada, es redundante crear una nueva clase para ello.

evaanngiil commented on src/GestorListaRegalos.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.HashMap;
+
+public class GestorListaRegalos { 
+    private HashMap<String, ListaRegalos> listas;
```
¿Qué representa la clave del diccionario?

pablotl0 replied:
hace referencia al destinatario de la lista de regalos

pablotl0 committed:
#23 el presupuesto se modela como un valor total dentro de la lista de regalos con el objetivo de simplififcar el modelo actual al no requerir mas funcionalidades el presupuesto

pablotl0 committed:
#13 se elimina la clase que modela el elemento de la lista porque de esta funcionalidad se encargará el regalo segun [https://github.com/evaanngiil/WishfulGiving/pull/18#discussion_r1867952954]

pablotl0 committed:
#14 modifico la clase Regalo para que este tenga un atributo de prioridad.  La prioridad facilitará la planificación y organización, permitiendo que los regalos más relevantes o urgentes sean identificados y tratados primero. Como se ha comentado en https://github.com/evaanngiil/WishfulGiving/pull/18#discussion_r1867952954

pablotl0 commented:
@evaanngiil ya he corregido los cambios

evaanngiil commented on src/Regalo.java:
```diff
@@ -0,0 +1,20 @@
+    private final String nombre;
+    private final double precio;
+    private final int prioridad;
+
```
Creo que el atributo `nombre` no incluiría información completa del regalo a hacer. Si bien creo que que lo más adecuando sería un `título` genérico y una `descripción`donde se diga de forma más específica las condiciones y detalles del regalo (como la talla, el color, el tamaño, etc)

pablotl0 committed:
#14 modifico la clase Regalo para que este tenga un atributo de descripción, relevante a la hora de especificar información adicional que describa mejor el regalo. Ademas cambio el atributo nombre por titulo, todo esto con el objetivo de mejorar la representacion del problema

pablotl0 commented:
@evaanngiil Ya está actualizado

evaanngiil reviewed (approved):
Ahora sí creo que cumple con el producto mínimamente viable requerido en #4

pablotl0 commented:
@JJ  Listo para revisión

JJ reviewed (approved):
:+1:

pablotl0 committed:
#11 actualizo la entidad del proyecto
